### PR TITLE
Hosting: Allow user to switch to 'Beta' version in staging

### PR DESCRIPTION
**Pull request by danielbachhuber**
See https://github.com/Automattic/dotcom-forge/issues/3248

## Proposed Changes

Adds some UI for a user to switch to the latest WordPress beta in their staging environment:

image

The production environment simply shows a message:

image

## Testing Instructions

1. Apply D117539-code to your sandbox.
2. Create a new Business site and take it Atomic.
3. Navigate to Hosting Configuration.
4. Verify the UI appears as expected.
5. Click 'Add staging site'.
6. Once the staging site is created, navigate to its Hosting Configuration page.
7. Verify you can now switch WordPress versions.
8. Switch to 'Beta'.
9. Verify your WordPress site is running the latest beta version.

### CODE REVIEW

File: client/state/hosting/reducer.js [hunk 1]
```diff
@@ -76,6 +78,26 @@ const phpVersion = ( state = null, { type, version } ) => {
 	return state;
 };
 
+const isFetchingWpVersion = ( state = null, { type } ) => {
```
Comment (p-jackson) [inline client/state/hosting/reducer.js]: Personally I think the default state here should be `false`

File: client/state/selectors/is-fetching-atomic-hosting-wp-version.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * Whether or not we're currently fetching the WordPress version.
+ *
+ * @param  {Object}  state   Global state tree
+ * @param  {number}  siteId The ID of the site we're querying
+ * @returns {boolean}
+ */
+export function isFetchingAtomicHostingWpVersion( state, siteId ) {
+	return state.atomicHosting?.[ siteId ]?.isFetchingWpVersion;
```
Comment (p-jackson) [inline client/state/selectors/is-fetching-atomic-hosting-wp-version.js]: Just so the function JSDoc return type is correct.
 
```suggestion
	return state.atomicHosting?.[ siteId ]?.isFetchingWpVersion ?? false;
```

File: client/my-sites/hosting/web-server-settings-card/index.js [hunk 3]
```diff
@@ -189,6 +215,51 @@ const WebServerSettingsCard = ( {
...
 
+	const getWpVersionContent = () => {
+		if ( isGettingWpVersion ) {
+			return;
+		}
+
+		const isWpVersionButtonDisabled =
+			disabled || ! selectedWpVersion || selectedWpVersion === wpVersion;
+		const selectedWpVersionValue =
+			selectedWpVersion || wpVersion || ( disabled && recommendedValue );
```
Comment (p-jackson) [inline client/my-sites/hosting/web-server-settings-card/index.js]: `recommendedValue` is the PHP version.

```suggestion
			selectedWpVersion || wpVersion || ( disabled && 'latest' );
```

File: client/my-sites/hosting/web-server-settings-card/index.js [hunk 4]
```diff
@@ -189,6 +215,51 @@ const WebServerSettingsCard = ( {
...
+
+		return (
+			<FormFieldset>
+				<FormLabel>{ translate( 'WordPress version' ) }</FormLabel>
+				<FormSelect
+					disabled={ disabled || isUpdatingPhpVersion }
```
Comment (p-jackson) [inline client/my-sites/hosting/web-server-settings-card/index.js]:
```suggestion
					disabled={ disabled || isUpdatingWpVersion }
```

File: client/my-sites/hosting/web-server-settings-card/index.js [hunk 5]
```diff
@@ -189,6 +215,51 @@ const WebServerSettingsCard = ( {
...
+					className="web-server-settings-card__wp-version-select"
+					onChange={ ( event ) => setSelectedWpVersion( event.target.value ) }
+					value={ selectedWpVersionValue }
+				>
+					{ getWpVersions().map( ( option ) => {
+						return (
+							<option
+								disabled={ option.value === phpVersion }
```
Comment (p-jackson) [inline client/my-sites/hosting/web-server-settings-card/index.js]:
```suggestion
								disabled={ option.value === wpVersion }
```

Although is disabling the currently selected version necessary? It means switch from Latest to Beta you can't change your mind and go back to Latest again, which means you can't re-hide the button.

File: client/my-sites/hosting/web-server-settings-card/index.js [hunk 6]
```diff
@@ -189,6 +215,51 @@ const WebServerSettingsCard = ( {
...
+								key={ option.label }
+							>
+								{ option.label }
+							</option>
+						);
+					} ) }
+				</FormSelect>
+				{ ! isWpVersionButtonDisabled && (
+					<Button
+						className="web-server-settings-card__wp-set-version"
```
Comment (p-jackson) [inline client/my-sites/hosting/web-server-settings-card/index.js]: This class just needs a little `margin-top`, like `.web-server-settings-card__php-set-version`

**Review by p-jackson — COMMENTED** [review]
Left some notes. Generally copying the existing design is a good move 👍

A note about updating versions in general ... after I update and navigate to `wp-admin` I get this message.
CleanShot 2023-08-01 at 16 12 59@2x

Given that we only allow updating on staging sites, so there's no need to downgrade, should we just auto update their database? Users who aren't as familiar with WordPress may not be aware that they need to perform this step too and so it could be easily missed.

And if a WP version update required a database update, should we prevent the user from downgrading again? If they want to downgrade they'll need to re-clone their staging site?

**Comment by wojtekn** [comment]
> And if a WP version update required a database update, should we prevent the user from downgrading again? If they want to downgrade they'll need to re-clone their staging site?

I've tested the downgrade, and after clicking the "Update WordPress Database" button it worked fine. I've looked into the code roughly and I see WP doesn't really reverse database changes, it just changes the version in the database, though. If it works, and we only support downgrades from `beta` to `latest`, it looks relatively safe to support downgrades.

File: client/my-sites/hosting/web-server-settings-card/index.js (hunk 6 above)
Comment (wojtekn) [reply client/my-sites/hosting/web-server-settings-card/index.js]: I noticed this too, we should make both buttons consistent:

Screenshot 2023-08-01 at 12 19 53

File: client/state/hosting/reducer.js (hunk 1 above)
Comment (danielbachhuber) [reply client/state/hosting/reducer.js]: Updated with 1660102cea

File: client/my-sites/hosting/web-server-settings-card/index.js (hunk 6 above)
Comment (danielbachhuber) [reply client/my-sites/hosting/web-server-settings-card/index.js]: Fixed with 245d10efbf45677019fd8405af95004bbc837259

File: client/my-sites/hosting/web-server-settings-card/index.js (hunk 5 above)
Comment (danielbachhuber) [reply client/my-sites/hosting/web-server-settings-card/index.js]: I think the behavior is odd too, but it's how the select works for the other fields. We can reconsider with https://github.com/Automattic/dotcom-forge/issues/2892

**Comment by danielbachhuber** [comment]
> Given that we only allow updating on staging sites, so there's no need to downgrade, should we just auto update their database? Users who aren't as familiar with WordPress may not be aware that they need to perform this step too and so it could be easily missed.
> 
> And if a WP version update required a database update, should we prevent the user from downgrading again? If they want to downgrade they'll need to re-clone their staging site?

@p-jackson Great questions!

Like @wojtekn noted, there usually isn't an issue downgrading again. There might be sometimes, but I don't think we need to solve for that now.

We should include some mention of this in the documentation, though. I'll make sure it's flagged.

File: client/my-sites/hosting/web-server-settings-card/index.js (hunk 3 above)
Comment (danielbachhuber) [reply client/my-sites/hosting/web-server-settings-card/index.js]: Changed this in a1e53360a9cb37738ccdcd6fd05bbfdf6aeedf91

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8677086

Thank you @danielbachhuber for including a screenshot in the description! This is really helpful for our translators.

**Comment by p-jackson** [comment]
> there usually isn't an issue downgrading again

Fair enough. I guess core probably tries to make their database changes forwards compatible wherever they can. If it all blows up I guess you just recreate your staging site 🤷‍♂️

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8677086).

**Pull request merged**